keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1316 @@
+node unschedulable. We will also change the implementation to skip applications that support
+workload migration. This will be detected by observing a `EvacuationRequest` condition on the pod
+and the subsequent appearance of `EvacuationInitiated` condition within a reasonable timeframe (3m).
+At first only deployments with `.spec.strategy.rollingUpdate.maxSurge` value are expected to
+respond to this request. If the cluster doesn't support the NodeMaintenance API, kubectl will
+perform the node drain in a backwards compatible way.
+
+`kubectl cordon` and `kubectl uncordon` commands will be enhanced with a warning that will warn
+the user if a node is made un/schedulable, and it collides with an existing NodeMaintenance object.
+As a consequence the node maintenance controller will reconcile the node back to the old value.
+
+### NodeMaintenance API
+
+NodeMaintenance objects serve as an intent to remove or migrate pods from a set of nodes. We will
+include Cordon and Drain toggles to support the following states/stages of the maintenance:
+1. Planning: this is to let the users know that maintenance will be performed on a particular set
+   of nodes in the future. Configured with `.spec.stage=Idle`.
+   2. Cordon: stop accepting (scheduling) new pods. Configured with `.spec.stage=Cordon`.
+3. Drain: gives an intent to drain all selected nodes by setting an `EvacuationRequest`
+   condition with `Reason="NodeMaintenance"` on the node's pods. Configured with
+   `.spec.stage=Drain`.
+4. Drain Complete: all targeted pods have been drained from all the selected nodes. The nodes can
+   be upgraded, restarted, or shut down. The configuration is still kept at `.spec.stage=Drain` and
+   `Drained` condition is set to `"True"` on the node maintenance object.
+5. Maintenance Complete: make the nodes schedulable again once the node maintenance is done.
+   Configured with `.spec.stage=Complete`.
+
+```golang
+
+// +enum
+type NodeMaintenanceStage string
+
+const (
+// Idle does not interact with the cluster.
+Idle NodeMaintenanceStage = "Idle"
+// Cordon cordons all selected nodes by making them unschedulable.
+Cordon NodeMaintenanceStage = "Cordon"
+// Drain:
+// 1. Cordons all selected nodes by making them unschedulable.
+// 2. Gives an intent to drain all selected nodes by setting an EvacuationRequest condition on the
+//    node's pods.
+Drain NodeMaintenanceStage = "Drain"
+// Complete:
+// 1. Removes EvacuationRequest on all the pods targeted by this NodeMaintenance.
+// 2. Uncordons all selected nodes by making them schedulable again, unless there is not another
+//    maintenance in progress.
+Complete NodeMaintenanceStage = "Complete"
+)
+
+type NodeMaintenance struct {
+    ...
+    Spec NodeMaintenanceSpec
+    Status NodeMaintenanceStatus
+}
+
+type NodeMaintenanceSpec struct {
+    // +required
+    NodeSelector *v1.NodeSelector
+
+    // The order of the stages is Idle -> Cordon -> Drain -> Complete.
```

Comment:
Post-alpha, we might want to support Idle → Cancelled and Idle → Cordon → Cancelled, to record that we never triggered the drain.

Reply:
I have to update the documentation here, but it should be similar to what can be found in the `Supported Stage Transitions` section.

A good idea: it would be good to record the passed stages in the status.